network/local.go

```diff
@@ -0,0 +1,459 @@
+package network
+
+import (
+	"errors"
+	"fmt"
+	"golang.org/x/net/context"
+	"reflect"
+	"sync"
+	"time"
+)
+
+// NewLocalRouter returns a fresh router which uses local connections. It takes
+// the default context.
+func NewLocalRouter(sid *ServerIdentity) (*Router, error) {
+	return NewLocalRouterWithContext(defaultLocalContext, sid)
+}
+
+// NewLocalRouterWithContext is the same as NewLocalRouter but takes a specific
+// LocalContext. This is useful to run parallel different local overlays.
+func NewLocalRouterWithContext(ctx *LocalContext, sid *ServerIdentity) (*Router, error) {
+	h, err := NewLocalHostWithContext(ctx, sid.Address)
+	if err != nil {
+		return nil, err
+	}
+	return NewRouter(sid, h), nil
+}
+
+// LocalContext keeps reference to all opened local connections
+// It also keeps tracks of who is "listening", so it's possible to mimics
+// Conn & Listener.
+type LocalContext struct {
+	// queues maps a remote endpoint to its packet queue. It's the main
+	//stucture used to communicate.
+	queues map[endpoint]*connQueue
+	sync.Mutex
+	listening map[Address]func(Conn)
+
+	baseUID uint64
+}
+
+// NewLocalContext returns a fresh new context that can be used by LocalConn,
+// LocalListener & LocalHost.
+func NewLocalContext() *LocalContext {
+	return &LocalContext{
+		queues:    make(map[endpoint]*connQueue),
+		listening: make(map[Address]func(Conn)),
+	}
+}
+
+var defaultLocalContext = NewLocalContext()
```

Comment:
> This sounds like 5-10% aren't fine?

By fine you mean they can't be run in parallel then yes. But it's not the only global variable we have in our sda tests (registered services, registered protocols etc). 
Seeing all kinds of reactions around `LocalContext`, I propose something:
* I make *all* sda tests using `LocalTest` so they can't all be ran in parallel.
* I make all functions of `LocalContext` **unexported** so we can't do anything with a `LocalContext`. I should already have done that anyway.
* I keep the global var as for me it's does not really make senses to force any user to create and pass around an explicit `LocalContext` while he does not need to. There's no un-mutex-protected access and since **all methods will be unexported**, he can't do much with it anyway.